library/bignum_mod.c

```diff
@@ -191,6 +191,95 @@ int mbedtls_mpi_mod_sub( mbedtls_mpi_mod_residue *X,
 
     return( 0 );
 }
+
+static int mbedtls_mpi_mod_inv_mont( mbedtls_mpi_mod_residue *X,
+                                     const mbedtls_mpi_mod_residue *A,
+                                     const mbedtls_mpi_mod_modulus *N,
+                                     mbedtls_mpi_uint *working_memory )
+{
+    /* Input already in Montgomery form, so there's little to do */
+    mbedtls_mpi_mod_raw_inv_prime( X->p, A->p,
+                                   N->p, N->limbs,
+                                   N->rep.mont.rr,
+                                   working_memory );
+    return( 0 );
+}
+
+static int mbedtls_mpi_mod_inv_non_mont( mbedtls_mpi_mod_residue *X,
+                                         const mbedtls_mpi_mod_residue *A,
+                                         const mbedtls_mpi_mod_modulus *N,
+                                         mbedtls_mpi_uint *working_memory )
+{
+    /* Need to convert input into Montgomery form */
+
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+
+    mbedtls_mpi_mod_modulus Nmont;
+    mbedtls_mpi_mod_modulus_init( &Nmont );
+
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mod_modulus_setup( &Nmont, N->p, N->limbs,
+                                         MBEDTLS_MPI_MOD_REP_MONTGOMERY ) );
+
+    /* We'll use X->p to hold the Montgomery form of the input A->p */
+    mbedtls_mpi_core_to_mont_rep( X->p, A->p, Nmont.p, Nmont.limbs,
```

Comment:
Any reason we are not invoking the `mbedtls_mpi_core_to_mont_rep()`? I assume that the main reason behind that design is memory ownership, and in particular who owns the memory pointed by `*working_memory`

Given this use case I would rather we modify `mbedtls_mpi_mod_raw_to_mont_rep` to now allocate memory and have a high level mod_owned version to do so.

### High level mod_to_mont_rep (owning the memory pointed by `*T` )
~~~~~
int mbedtls_mpi_mod_to_mont_rep( mbedtls_mpi_uint *X,
                                     const mbedtls_mpi_mod_modulus *m)
{
    mbedtls_mpi_uint *T;
    const size_t t_limbs = m->limbs * 2 + 1;

    if( ( T = (mbedtls_mpi_uint *) mbedtls_calloc( t_limbs, ciL ) ) == NULL )
        return( MBEDTLS_ERR_MPI_ALLOC_FAILED );
    
    mbedtls_mpi_mod_raw_to_mont_rep(X, m, T)
    mbedtls_mpi_core_montmul( X, X, m->rep.mont.rr, m->limbs, m->p, m->limbs,
                              m->rep.mont.mm, T );

    mbedtls_platform_zeroize( T, t_limbs * ciL );
    mbedtls_free( T );
    return( 0 );
}
~~~~~~


### Mid Level mod_to_mont_rep (Abstracting the core to  montumul)
~~~~~~
int mbedtls_mpi_mod_raw_to_mont_rep( mbedtls_mpi_uint *X,
                                     const mbedtls_mpi_mod_modulus *m,
                                     mbedtls_mpi_uint *T)
{
    const size_t t_limbs = m->limbs * 2 + 1;

    mbedtls_mpi_core_montmul( X, X, m->rep.mont.rr, m->limbs, m->p, m->limbs,
                              m->rep.mont.mm, T );

    return( 0 );
}
~~~~~

Reply:
The main difference is that I wanted a version of Montgomery conversion that wasn't in-place.

Since the conversions we currently have are in mod_raw, it made sense to put the new ones in core - I assume that the design and location of the current ones is as per spec, so I'm not going to move them around - certainly not in this PR.

Since core already understands stuff that is in Montgomery form only, it doesn't seem to me to be a problem to have a thin pair of conversion functions there.

If we want to rewrite the montgomery conversion in mod_raw to not allocate, we can, but that won't help me especially here - I'd still have the in-place issue. And I presume having the mod_raw version allocate was part of its origin design.

Reply:
I have created a TD item to track this , and allow the team discuss on the design #6815. I will resolve this since it is now out of scope.